src/utils/constants.hxx

```diff
@@ -102,19 +107,47 @@
         return (ptrdiff_t)(&_sym_##name);                                      \
     }
 
+/// Defines the default value of a constant. Use this in a single .cxx file and
+/// make sure NOT to include the header that has the respective DECLARE_CONST
+/// macros. Best not to incude anything at all.
+///
+/// @param name name of the constant.
+/// @param value is what the default value should be.
 #define DEFAULT_CONST(name, value)                                             \
     typedef signed char                                                        \
     _do_not_add_declare_and_default_const_to_the_same_file_for_##name;         \
     asm(".global _sym_" #name " \n");                                          \
     asm(".weak _sym_" #name " \n");                                            \
     asm(".set _sym_" #name ", " #value " \n");
 
+/// Overrides the value of a constant. Use this in a single .cxx file (usually
+/// main.cxx).
+///
+/// @param name name of the constant.
+/// @param value is what the actual value should be.
 #define OVERRIDE_CONST(name, value)                                            \
     asm(".global _sym_" #name " \n");                                          \
     asm(".set _sym_" #name ", " #value " \n");
 
 #endif // native C
 
+/// Defines the default value of a constant using an externally defined value
+/// by using deferred expansion to work around pre-compiler limitations. Like
+/// @ref DEFAULT_CONST use this in a single .cxx file and make sure NOT to
+/// include the header that has the respective DECLARE_CONST macro.
+///
+/// @param var name of the constant.
+/// @param value is what the actual value should be.
+#define DEFAULT_CONST_DEFERRED(var, value) DEFAULT_CONST(var, value)
```

Comment:
do we really need this complexity in the base header? It's somewhat unclear to me when and why would you want to use it ("deferred expansion" is not a particularly descriptive comment on a macro called "_DEFERRED").

Could this complexity move to the location where you actually need it?

Can you put a single DEFER macro into the value instead of duplicating every macro, like OVERRIDE_CONST(foo, DEFER(bar));

Reply:
The usage of this is in expanding of an externally defined value, such as:
```
OVERRIDE_CONST_DEFERRED(gridconnect_buffer_size, CONFIG_LWIP_TCP_MSS);
```

I'll see if I can do this with `OVERRIDE_CONST(gridconnect_buffer_size, DEFER(CONFIG_LWIP_TCP_MSS));` instead. If that works I'll simplify this to just have `DEFER(value)` instead.

Reply:
another trick I've seen is to add an EMPTY() macro (which expands to nothing) and have this be present in the expansion. If that works, we could have that as the default behavior, i.e. modify DEFAULT_CONST and OVERRIDE_CONST directly.

Reply:
That sounds like a better option than what I've implemented here. How does `CONSTANT_DEFER` sound for the extra macro defined in constants.hxx and used in the `DEFAULT_CONST` and `OVERRIDE_CONST` macros?

Reply:
In testing this I found a problem with the inline expansion of `DEFER(value)`, specifically when `NEED_SIMPLE_CONST` is not defined. The `asm(...)` lines are using concatenation to inject the value into the generated asm code and that will prevent this from working as expected.

One option to fix this would be ensure that NEED_SIMPLE_CONST is used by the ESP32, another would be use it for all platforms (likely not ideal).

Thoughts?

Reply:
I definitely don't want to use simple const for all platforms. Have you tried EMPTY() ?

Reply:
The `asm(...)` expansion fails unfortunately:
```
.../src/utils/constants.cpp.obj.d -o esp-idf/OpenMRNLite/CMakeFiles/__idf_OpenMRNLite.dir/src/utils/constants.cpp.obj -c ../components/OpenMRNLite/src/utils/constants.cpp
/tmp/ccUyDoir.s: Assembler messages:
/tmp/ccUyDoir.s:129: Error: junk at end of line, first unrecognized character is `('
ninja: build stopped: subcommand failed.
```
The code used was:
```
DEFAULT_CONST(arduino_openmrn_task_priority, EMPTY(ESP_TASK_TCPIP_PRIO - 1));
```

Reply:
One more option, I can rename the two current ones to `DEFAULT_CONST_INLINE` and `OVERRIDE_CONST_INLINE` and rename the deferred variant to the current names. This way the ability to override with another compile time constants is available and using the two types of constant value definitions is also preserved.